pkg/services/k8s/authnz/service.go

```diff
@@ -0,0 +1,141 @@
+package authnz
+
+import (
+	"context"
+	"fmt"
+
+	"net/http"
+	"strconv"
+
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/api/routing"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models/roletype"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/authn/clients"
+	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/web"
+	authnV1 "k8s.io/api/authentication/v1"
+	authzV1 "k8s.io/api/authorization/v1"
+)
+
+const GrafanaAdminK8sUser = "gl-admin"
+
+type K8sAuthnzAPIImpl struct {
+	RouteRegister routing.RouteRegister
+	AccessControl accesscontrol.AccessControl
+	ApiKey        *clients.APIKey
+	Log           log.Logger
+}
+
+func ProvideService(
+	rr routing.RouteRegister,
+	apikeyService apikey.Service,
+	userService user.Service,
+	ac accesscontrol.AccessControl,
+) *K8sAuthnzAPIImpl {
+	k8sAuthnzAPI := &K8sAuthnzAPIImpl{
+		RouteRegister: rr,
+		AccessControl: ac,
+		ApiKey:        clients.ProvideAPIKey(apikeyService, userService),
```

Comment:
Suggestion: Instead of constructing the api client here you should inject authn.Service and perform authentication through that service. That one will test several authentication clients but only api key client would be able to authenticate request coming through here.

Reply:
It also does feel like overkill to use `authn.Service` in this scenario knowing that we already have the pre-determined client we are aware we should use. So just throwing that out there to see if the token authentication method exposed by that service could potentially provide a wrapped form which does record the key use data but is more similar to the `authenticate` method I am already using.

Reply:
Okay, I see why the `authn.Service` is the recommended way to go. It performs these post hooks for all clients and is definitely more elegant.

```
	s.RegisterPostAuthHook(userSyncService.SyncUserHook, 10)
	s.RegisterPostAuthHook(userSyncService.EnableDisabledUserHook, 20)
	s.RegisterPostAuthHook(orgUserSyncService.SyncOrgRolesHook, 30)
	s.RegisterPostAuthHook(userSyncService.SyncLastSeenHook, 40)
```

Reply:
We decided to go forward with `authn.Service` and kept the authorization header hack. In future, we can always revisit if the hack is the long term solution we want.